### PR TITLE
compiler: fixes calling convention violation.

### PR DESCRIPTION
**Pull request by mathetake**
If the call_indirect's offset register is the same as the calling-convention
reserved register, the function call target address resulted in the wrong 
address and therefore segfault.

Fixes partially #526 

Signed-off-by: Takeshi Yoneda <takeshi@tetrate.io>

### CODE REVIEW

**Comment by mathetake** [comment]
will backfill the regression tests as a toehold!

File: internal/engine/compiler/impl_amd64.go [hunk 1]
```diff
@@ -4420,14 +4421,19 @@ func (c *amd64Compiler) compileCallFunctionImpl(index wasm.Index, codeAddressReg
 	// So we increment the call frame stack pointer.
 	c.assembler.CompileNoneToMemory(amd64.INCQ, amd64ReservedRegisterForCallEngine, callEngineGlobalContextCallFrameStackPointerOffset)
 
+	if amd64CallingConventionModuleInstanceAddressRegister == targetFunctionAddressRegister {
```
Comment (mathetake) [inline internal/engine/compiler/impl_amd64.go:4424]: This is the key, others are just styles

**Comment by mathetake** [comment]
ok fixed arm64 as well as added unit tests!

File: internal/engine/compiler/impl_amd64.go [hunk 2]
```diff
@@ -4420,14 +4421,22 @@ func (c *amd64Compiler) compileCallFunctionImpl(index wasm.Index, codeAddressReg
...
+		// This case we must move the value on targetFunctionAddressRegister to another register, otherwise
+		// the address (jump target below) will be modified and result in segfault.
+		// See $526.
```
Comment (anuraaga) [inline internal/engine/compiler/impl_amd64.go:4427]: `#526` or consider actual url for this and others

File: internal/engine/compiler/impl_amd64.go (hunk 1 above)
Comment (anuraaga) [reply internal/engine/compiler/impl_amd64.go:4424]: Just wondering, is the reason to only mov in this branch an optimization to save a register for other use?

**Review by anuraaga — APPROVED** [review]

File: internal/engine/compiler/impl_amd64.go (hunk 1 above)
Comment (mathetake) [reply internal/engine/compiler/impl_amd64.go:4424]: yeah, optimization as not always end up in this situation, and the less instructions the better in general!

**Pull request merged**